src/apiserver/api_service_impl.h

```diff
@@ -0,0 +1,287 @@
+                                   std::shared_ptr<fedb::sdk::SQLRequestRow> row);
+    template <typename T>
+    static bool AppendJsonValue(const butil::rapidjson::Value& v, hybridse::sdk::DataType type, bool is_not_null,
+                                T row);
+
+    static SimpleSchema TransToSimpleSchema(const hybridse::sdk::Schema* schema) {
+        SimpleSchema ss;
+        for (int i = 0; i < schema->GetColumnCnt(); ++i) {
+            ss.emplace_back(schema->GetColumnName(i), schema->GetColumnType(i), schema->IsConstant(i),
+                            schema->IsColumnNotNull(i));
+        }
+        return ss;
+    }
+
+ private:
+    std::unique_ptr<sdk::SQLRouter> sql_router_;
+    InterfaceProvider provider_;
+};
+
+struct Column {
+    std::string name;
+    hybridse::sdk::DataType type = hybridse::sdk::kTypeUnknow;
+    bool is_constant = false;
+    bool is_null = false;
+    Column() = default;
+    Column(std::string n, hybridse::sdk::DataType t, bool c, bool null)
+        : name(std::move(n)), type(t), is_constant(c), is_null(null) {}
+};
+
+template <typename Archiver>
+Archiver& operator&(Archiver& ar, Column& s) {  // NOLINT
+    if (ar.IsReader) {
+        LOG(WARNING) << "unsupported now, reason: the format is '{name}:{type}', we can't read the variable name.";
+        return ar;
+    }
+    ar.StartObject();
+    ar.Member(s.name.c_str()) & hybridse::sdk::DataTypeName(s.type);
+    return ar.EndObject();
+}
+
+struct PutResp {
+    PutResp() = default;
+    int code = 0;
+    std::string msg = "ok";
+};
+
+template <typename Archiver>
+Archiver& operator&(Archiver& ar, PutResp& s) {  // NOLINT
+    ar.StartObject();
+    ar.Member("code") & s.code;
+    ar.Member("msg") & s.msg;
+    return ar.EndObject();
+}
+
+struct ExecSPResp {
+    ExecSPResp() = default;
+    int code = 0;
+    std::string msg = "ok";
+    struct Data {
+        SimpleSchema schema;
+        std::shared_ptr<hybridse::sdk::ResultSet> rs;
+    };
+    Data data;
+};
+
+template <typename Archiver, typename Type>
+void WriteArray(Archiver& ar, const std::string& name, std::vector<Type>& vec) {  // NOLINT
+    ar.Member(name.c_str());
+    size_t count = vec.size();
+    ar.StartArray();
+    for (size_t i = 0; i < count; i++) {
+        ar& vec[i];
+    }
+    ar.EndArray();
+}
+
+template <typename Archiver>
+void WriteValue(Archiver& ar, std::shared_ptr<hybridse::sdk::ResultSet> rs, SimpleSchema& schema,  // NOLINT
+                int i) {
+    if (rs->IsNULL(i)) {
+        if (!schema[i].is_null) {
+            LOG(ERROR) << "Value in " << schema[i].name << " is null but it can't be null";
+        }
+        ar.SetNull();
+        return;
+    }
+    switch (schema[i].type) {
+        case hybridse::sdk::kTypeInt32: {
+            ar & rs->GetInt32Unsafe(i);
+            break;
+        }
+        case hybridse::sdk::kTypeInt64: {
+            ar & rs->GetInt64Unsafe(i);
```

Comment:
use "<<"

Reply:
It's for both writer & reader, can't use "<<"